overlord/restart/restart_test.go

```diff
@@ -428,10 +505,251 @@ func (s *restartSuite) TestPendingForSystemRestart(c *C) {
+
+	err = restart.TaskWaitForRestart(t1)
+	c.Assert(err, IsNil)
+
+	c.Check(t1.Log(), HasLen, 1)
+	c.Check(t1.Log()[0], Matches, ".* Task set to wait until a system restart allows to continue")
+
+	var waitBootID string
+	if err := t1.Get("wait-for-system-restart-from-boot-id", &waitBootID); !errors.Is(err, state.ErrNoState) {
+		c.Check(err, IsNil)
+	}
+	c.Check(waitBootID, Equals, "boot-id-1")
+	c.Check(t1.Status(), Equals, state.WaitStatus)
+	c.Check(t1.WaitedStatus(), Equals, state.DoStatus)
+
+	c.Check(t1.Log(), HasLen, 1)
+	c.Check(t1.Log()[0], Matches, ".* Task set to wait until a system restart allows to continue")
+}
+
+func (s *restartSuite) TestTaskWaitForRestartUndoClassic(c *C) {
+	restore := release.MockOnClassic(true)
+	defer restore()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("test", "...")
+	t1 := st.NewTask("foo", "...")
+	chg.AddTask(t1)
+
+	t1.SetStatus(state.UndoingStatus)
+
+	err := restart.TaskWaitForRestart(t1)
+	c.Assert(err, IsNil)
+
+	c.Check(t1.Log(), HasLen, 1)
+	c.Check(t1.Log()[0], Matches, ".* Skipped automatic system restart on classic system when undoing changes back to previous state")
+}
+
+func (s *restartSuite) TestTaskWaitForRestartUndoCore(c *C) {
+	restore := release.MockOnClassic(false)
+	defer restore()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	_, err := restart.Manager(st, "boot-id-1", nil)
+	c.Assert(err, IsNil)
+
+	chg := st.NewChange("test", "...")
+	t1 := st.NewTask("foo", "...")
+	chg.AddTask(t1)
+
+	t1.SetStatus(state.UndoingStatus)
+
+	err = restart.TaskWaitForRestart(t1)
+	c.Assert(err, IsNil)
+
+	c.Check(t1.Log(), HasLen, 1)
+	c.Check(t1.Log()[0], Matches, ".* Task set to wait until a system restart allows to continue")
+
+	var waitBootID string
+	if err := t1.Get("wait-for-system-restart-from-boot-id", &waitBootID); !errors.Is(err, state.ErrNoState) {
+		c.Check(err, IsNil)
+	}
+	c.Check(waitBootID, Equals, "boot-id-1")
+	c.Check(t1.Status(), Equals, state.WaitStatus)
+	c.Check(t1.WaitedStatus(), Equals, state.UndoStatus)
+
+	c.Check(t1.Log(), HasLen, 1)
+	c.Check(t1.Log()[0], Matches, ".* Task set to wait until a system restart allows to continue")
+}
+
+func (s *restartSuite) TestTaskWaitForRestartInvalid(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("test", "...")
+	t1 := st.NewTask("foo", "...")
+	chg.AddTask(t1)
+
+	err := restart.TaskWaitForRestart(t1)
+	c.Assert(err, ErrorMatches, `internal error: only tasks currently in progress \(doing/undoing\) are supported`)
+}
+
+func (s *restartParametersSuite) TestFinishTaskWithRestart2Done(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	_, err := restart.Manager(st, "boot-id-1", nil)
+	c.Assert(err, IsNil)
+
+	chg := st.NewChange("test", "...")
+	t := st.NewTask("waiting", "...")
+	chg.AddTask(t)
+
+	err = restart.FinishTaskWithRestart2(t, "some-snap", state.DoneStatus, restart.RestartSystemNow, nil)
```

Comment:
I think that some test for `FinishTaskWithRestart2` where `rebootInfo` is not nil is missing

Reply:
Good point!